Add Failure Details to every NodesResponse

pickypg opened the pull request:
Most of the current implementations of `BaseNodesResponse` (plural Nodes) ignore `FailedNodeException`s.

- This adds a helper function to do the grouping to `TransportNodesAction`
- Requires a non-null array of `FailedNodeException`s within the `BaseNodesResponse` constructor
- Reads/writes the array to `Streamable`
- Also adds `StreamInput` and `StreamOutput` methods for generically reading and writing arrays

/cc @jasontedor (for failures) @nik9000 (for StreamInput/Output helpers)

Closes #3740

pickypg commented on core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:
```diff
@@ -238,4 +269,49 @@ public void messageReceived(NodeRequest request, TransportChannel channel) throw
+         */
+        @SuppressWarnings("unchecked")
+        public ResponseContainer(Class<T> clazz, List<T> nodeResponses, List<FailedNodeException> failures) {
+            this.nodeResponses = nodeResponses.toArray((T[])Array.newInstance(clazz, nodeResponses.size()));
```
I don't really want to pass through `IntFunction<T[]> arrayCreator` for every single one, but I don't really like this variant either. Curious how others feel.

jasontedor replied:
I'd consider converting everything in `BaseNodesResponse` to a list instead of an array and then just passing the list around?

jasontedor replied:
> I'd consider converting everything in `BaseNodesResponse` to a list instead of an array and then just passing the list around?

@pickypg Can we do this?

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:
```diff
@@ -89,6 +92,34 @@ protected boolean transportCompress() {
+            }
+        }
+
+        return new ResponseContainer<>(expected, nodesResponses, failures);
```
I think that the `ResponseContainer` abstraction is completely unnecessary. You can just add an abstract protected method to `TransportNodesAction` that receives the collected responses and failures. Take a look at `TransportBroadcastByNodeAction` where a similar mechanism is employed (the grouping is handled in the base class, and then delegated to an abstract protected `newResponse` method to handle creating the correct response object).

jasontedor commented:
@pickypg Other than the `ToXContent` stuff that we discussed via another channel, I left a comment about the `ResponseContainer` abstraction; I do not think this is necessary and left a specific suggestion. I'll hold off on a thorough review until you have a chance to process those two things.

jasontedor commented:
Note that this will close #3740.

pickypg commented:
@jasontedor Updated. I force pushed as it touched a lot of the existing code. I do prefer the abstract method. I didn't think that _every_ instance was using it, but they were, so I definitely prefer it.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/nodes/BaseNodesResponse.java:
```diff
@@ -91,11 +105,36 @@ public TNodeResponse getAt(int position) {
+    /**
+     * Expected to be invoked by any {@code BaseNodesResponse} that implements {@code ToXContent} at the top level of the response object.
+     */
+    protected XContentBuilder toInnerXContent(XContentBuilder builder, Params params) throws IOException {
```
I don't like super methods that when they *are* overridden, they *must* be invoked. I also don't like that this method is on the base class and thus inherited by response objects that do not implement `ToXContent`, I think that is misleading. Lastly, I think that we should include total/successful/failed counts in the response.

My specific suggestion is to add a method to `RestActions` similar to `RestActions#buildBroadcastShardsHeader`. I think it can look something like this:

```java
    public static <TNodesResponse extends BaseNodeResponse> void buildNodesInfoHeader(
            final XContentBuilder builder,
            final ToXContent.Params params,
            final BaseNodesResponse<TNodesResponse> response) throws IOException {
        final TNodesResponse[] responses = response.getNodes();
        final FailedNodeException[] failures = response.failures();
        final int successful = responses != null ? responses.length : 0;
        final int failed = failures != null ? responses.length : 0;
        buildNodesInfoHeader(builder, params, successful + failed, successful, failed, failures);
    }

    public static void buildNodesInfoHeader(
            final XContentBuilder builder,
            final ToXContent.Params params,
            final int total,
            final int successful,
            final int failed,
            final FailedNodeException[] failedNodeExceptions) throws IOException {
        // nocommit: add a constant to Fields
        builder.startObject("_nodes");
        builder.field(Fields.TOTAL, total);
        builder.field(Fields.SUCCESSFUL, successful);
        builder.field(Fields.FAILED, failed);
        if (failedNodeExceptions != null && failedNodeExceptions.length > 0) {
            builder.startArray(Fields.FAILURES);
            for (FailedNodeException failedNodeException : failedNodeExceptions) {
                builder.startObject();
                failedNodeException.toXContent(builder, params);
                builder.endObject();
            }
            builder.endArray();
        }
        builder.endObject();
    }

    public static <TNodesResponse extends BaseNodesResponse & ToXContent> BytesRestResponse nodesInfoResponse(
            final XContentBuilder builder,
            final ToXContent.Params request,
            final TNodesResponse response) throws IOException {
        builder.startObject();
        RestActions.buildNodesInfoHeader(builder, request, response);
        response.toXContent(builder, request);
        builder.endObject();
        return new BytesRestResponse(RestStatus.OK, builder);
    }
```

And then `buildResponse` call sites can just look like `return RestActions.nodesInfoResponse(builder, request, response);`

pickypg replied:
I'm going to add the static method, but I do want to note that the method name is `toInnerXContent` rather than `toXContent`, so it's not overridden by any of the child implementations.

pickypg replied:
I really like the concept of the `"_nodes"`, but I'm not so sure about the name. Practically every `ToXContent` impl that I noticed had a `"nodes"` block. I'm a bit worried by how close those are.

How does that make you feel about `_nodes`? The problem is that I don't have a better name because that's the right one. :)

jasontedor replied:
> I'm going to add the static method, but I do want to note that the method name is toInnerXContent rather than toXContent, so it's not overridden by any of the child implementations.

Sure but it *can* be overridden, if it is overridden it must be called, and it has to be called by the `toXContent` methods on the inheriting classes that do implement `ToXContent` The typical pattern to address this is to make `toXContent` final and have it delegate to an abstract `doToXContent` inner method that the inheriting classes must override. But the reason that I do not like that solution here is because not all of the inheriting classes will implement `toXContent` so I do not think this method should be on the super class at all.

jasontedor replied:
> How does that make you feel about `_nodes`? The problem is that I don't have a better name because that's the right one. :)

Note that for the broadcast shards header in the response we use `_shards`. I think it should be the same here so `_nodes`?

pickypg replied:
> Note that for the broadcast shards header in the response we use _shards. I think it should be the same here so _nodes?

Convincing enough to me. I glanced around and a few of its uses did add a `"shards"` field (though not always at the same level).

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:
```diff
@@ -87,7 +99,49 @@ protected boolean transportCompress() {
+     * @throws NullPointerException if {@code nodesResponses} is {@code null}
+     * @see #newResponse(BaseNodesRequest, BaseNodeResponse[], FailedNodeException[])
+     */
+    @SuppressWarnings("unchecked") // generic array creation
```
If you separate out the array creation into its own line, the suppression can be applied to that line only rather than the entire method.

pickypg replied:
I think based on Nik's part, I'm going to change everything from an array to a `List`.

jasontedor commented:
@pickypg I think it looks great, I left a few more comments. 

> I force pushed as it touched a lot of the existing code.

😦

nik9000 commented on core/src/test/java/org/elasticsearch/common/io/stream/StreamInputTests.java:
```diff
@@ -0,0 +1,213 @@
+/**
+ * Tests {@link StreamInput}.
+ */
+public class StreamInputTests extends ESTestCase {
```
We've just been adding tests for this StreamInput and StreamOutput to BytesStreamsTests.

nik9000 replied:
Mockito isn't really required for testing these the way we do in BytesStreamsTests. which is always a plus.

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -386,6 +387,88 @@ public final Boolean readOptionalBoolean() throws IOException {
+    public <T> T[] readArray(IntFunction<T[]> arrayCreator, StreamInputFunction<T> converter) throws IOException {
+        int size = readVInt();
+
+        T[] array = arrayCreator.apply(size);
```
What about readList? I get that sometimes you have an array and not a list but this seems like overkill.

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -754,4 +837,23 @@ public static StreamInput wrap(byte[] bytes) {
+     * and it only allows {@link StreamInput} to be used as an argument.
+     */
+    @FunctionalInterface
+    public interface StreamInputFunction<R> {
```
StreamInputReader already does this.

pickypg replied:
As I was removing my added methods, I noticed that apparently `Writeable.Reader<R>` also adds this one (just FYI).

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -356,6 +356,90 @@ public void write(byte[] b, int off, int len) throws IOException {
+     * @param array The array to write
+     */
+    public <T extends Writeable> void writeArray(T[] array) throws IOException {
+        writeVInt(array.length);
```
Same comment on the other side - use writeList, probably. Either convert to a list when calling or just convert the object to storing a list rather than an array.

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -356,6 +356,90 @@ public void write(byte[] b, int off, int len) throws IOException {
+     * @see #writeArray(Writeable[])
+     */
+    public <T extends Streamable> void writeArray(T[] array) throws IOException {
+        writeVInt(array.length);
```
I'd prefer to replace implementations of Streamable with Writeable or some other hack like having Streamable extend Writeable. I'm not sure what the right hack is though.

pickypg replied:
I didn't want to have Streamable extend Writeable because I think it clouds the meaning of Writeable in those cases (its intent of immutability is suddenly lost when used with Streamable). It sounds like I'm going to shift to `List` from arrays anyway, and thereby drop these methods altogether.

jasontedor replied:
> It sounds like I'm going to shift to List from arrays anyway, and thereby drop these methods altogether.

+1, that should shrink the added code in this PR by a *lot*.

pickypg commented:
@jasontedor / @nik9000 

I changed it from `NodeResponse[]` to `List<NodeResponse>` and `FailedNodeException[]` to `List<FailedNodeException>`. I also dropped my changes from `StreamInput` and `StreamOutput` (and corresponding tests), but I did need to add a `<T extends Streamable> void writeStreamableList(List<T>)` (technically this could be `List<? extends Streamable>`, but I preferred matching the `Writable` variant) and `<T extends Streamable> List<T> readList(Supplier<T>))`.

Overall, I think it's better, but it naturally touches even more files. It took forever to get through the tests. :)

nik9000 commented:
> <T extends Streamable>

If you'd prefer to do `<? extends Streamable>` and `<? extends Writeable>` I'm fine with changing both. Or just doing yours with `?`. I suspect the `T` in the writeable one is because `Writeable` used to have a type parameter and that bullied people into adding the `T`. Maybe. Anyway, there isn't a reason to have it now that I can see.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -80,66 +81,46 @@ public ClusterStatsIndices getIndicesStats() {
-        nodesStats = new ClusterStatsNodes(in);
-        indicesStats = ClusterStatsIndices.readIndicesStats(in);
+        // it may be that the master switched on us while doing the operation. In this case the status may be null.
+        status = in.readBoolean() ? ClusterHealthStatus.fromValue(in.readByte()) : null;
```
It'd be "more normal" to declare this as `Writeable` and use `readOptionalWriteable` and `writeOptionalWriteable`. You've done plenty in this PR so it can wait though!

jasontedor replied:
As written this isn't symmetrical with the write method. I would prefer that it be written in a symmetrical way for ease of comparison.

pickypg replied:
Using the `writeOptionalWriteable` made it easy to make them symmetrical (had to make `ClusterHealthStatus` a `Writeable` though).

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -80,66 +81,46 @@ public ClusterStatsIndices getIndicesStats() {
     }
 
     static final class Fields {
         static final String NODES = "nodes";
```
I believe it is fine to just remove this class and use the strings. That is what we've been doing with new code. If we want field constants the Fields class is really nice. But I don't think the constants buy us much.

jasontedor replied:
I've also started removing them in places I touch. They cost us an extra class in the classloader that never gets unloaded.

pickypg replied:
Removed.

nik9000 commented on core/src/main/java/org/elasticsearch/action/support/nodes/BaseNodesResponse.java:
```diff
@@ -22,61 +22,82 @@
-    protected TNodeResponse[] nodes;
+    private List<FailedNodeException> failures;
+    /**
+     * Due to its generic nature, child classes are expected to read/write the nodes for {@code Streamable} contracts.
```
Thanks for leaving that comment!

nik9000 replied:
Now I wonder if you should make readFrom/writeTo final and make an abstract method that is just responsible for reading the response. That pattern was used by the query builders until they switched to constructor based reading and it seems fairly a appropriate. I don't think it should block the PR though. It isn't really important I think.

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/Writeable.java:
```diff
@@ -32,6 +32,7 @@
  * Prefer implementing this interface over implementing {@link Streamable} where possible. Lots of code depends on {@linkplain Streamable}
  * so this isn't always possible.
  */
+@FunctionalInterface
```
I don't think Writeable is intended to be used as a reference to a function. It is intended to be implemented. You can use it as a functional reference in rare cases but I don't think this interface is a good thing here.

nik9000 replied:
Sorry, this *annotation*.

pickypg replied:
> Sorry, this annotation.

It was clear what you meant. No worries. Removed.

nik9000 commented on core/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -57,24 +59,24 @@
      * An action that lists the relevant shard data that needs to be fetched.
      */
-    public interface List<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
+    public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
```
Yikes!

pickypg replied:
Just to be clear for me, you were referring to the need to rename this, right? :)

nik9000 replied:
Yes. Renaming that is a good thing.

pickypg commented:
@nik / @jasontedor 

I changed writeStreamableList and writeList to remove `T` as discussed. Also made other recommended changes.

nik9000 commented:
> I changed writeStreamableList and writeList to remove T as discussed. Also made other recommended changes.

Cool! I spent all morning sending emails and stuff so I'm itching to write code. I'll leave this open and have another look soonish. If @jasontedor gets to it before me fine by me!

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -80,76 +81,43 @@ public ClusterStatsIndices getIndicesStats() {
+        nodes = in.readList(ClusterStatsNodeResponse::readNodeResponse);
+
+        // built from nodes rather than from the stream directly
+        nodesStats = new ClusterStatsNodes(nodes);
```
Do these things need to be Writeable any more?

pickypg replied:
Nope. Removed it from both (worse: one was `Streamable` and one was `Writeable`)!

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/health/ClusterHealthStatus.java:
```diff
@@ -38,7 +44,24 @@ public byte value() {
+    }
+
+    /**
+     * Load a {@link ClusterHealthStatus}.
```
I usually just do
```
/**
 * Read from a stream.
 */
```

The input and output are relatively obvious. Same for the IOException. The IllegalArgumentException isn't, but honestly it isn't a big deal I think.

Also not a big deal, more a matter of my personal style preference which I don't want to force on anyone.

nik9000 commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -762,4 +785,5 @@ public static StreamInput wrap(byte[] bytes) {
     public static StreamInput wrap(byte[] bytes, int offset, int length) {
         return new InputStreamStreamInput(new ByteArrayInputStream(bytes, offset, length));
     }
+
```
Nit: why you add line break here? IDE do it for you?

jasontedor replied:
It's from a suggestion I made in an earlier pull request. The codebase is not consistent but there is definitely a tendency towards having a blank line between the opening of a class and its body, and its body and the closing of a class. It's not a requirement (but I wish that it was).

nik9000 replied:
I prefer it without the blank lines. No wonder we aren't consistent.

jasontedor replied:
It's not just you. Some classes even have a blank line between the opening and the body, but not the body of the class and the closing.

nik9000 replied:
They are just hedging their bets.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/action/support/RestActions.java:
```diff
@@ -64,18 +73,24 @@ public static long parseVersion(RestRequest request, long defaultVersion) {
     }
 
     static final class Fields {
+        static final String _NODES = "_nodes";
```
Nit: I'd just use the string rather than make a constant. We are slowly removing these objects.

pickypg replied:
One day Java will add a true `const` and we can actually use these types of constructs like we're trying to do here: ensure consistency without the bloat of adding yet another class / static reference.

But not this day.

I went ahead and removed them, but I do think that these things actually add value when the field is used more than once in the class, but it's not worth the cost (we really do have enough classes without adding noop ones like these).

jasontedor replied:
> I do think that these things actually add value when the field is used more than once in the class

Having a `static final String` constant is fine (if you refer to the constant, or refer to a literal string, it doesn't matter they will both end up in the constant pool for the class), it's adding an *extra* class to hold that is unnecessary.

pickypg replied:
Yeah, but you still get the wasted, additional reference to the constant pool (granted that's small potatoes, it's something that the compiler could auto-replace for us). Agreed!

jasontedor replied:
I'm not following you? The bytes for the string have to sit somewhere and they always have to be there. When the compiler needs to use a string from the constant pool, it just emits a special instruction `ldc` to load a reference to the string onto the stack.

pickypg replied:
The lingering `static` reference that must exist as a result.

jasontedor replied:
I'm still not following you? What's wrong with the `static` reference? Whether you use a constant string field (`static final String` assigned from a literal string) or just use literal strings in all the places that you would use the constant, the effect is the same: a single `String` is put in the constant pool and that constant is pushed onto the stack when needed using `ldc`. The java compiler effectively interns all literal strings, and this is the same effect as using a constant string field.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/action/support/RestActions.java:
```diff
@@ -93,6 +108,95 @@ public static void buildBroadcastShardsHeader(XContentBuilder builder, ToXConten
+    public static <NodesResponse extends BaseNodesResponse & ToXContent> BytesRestResponse nodesResponse(final XContentBuilder builder,
+                                                                                                         final Params params,
+                                                                                                         final NodesResponse response)
+        throws IOException {
```
Can you indent this so it doesn't look like method body?

nik9000 commented on core/src/main/java/org/elasticsearch/rest/action/support/RestActions.java:
```diff
@@ -93,6 +108,95 @@ public static void buildBroadcastShardsHeader(XContentBuilder builder, ToXConten
+    public static <NodeResponse extends BaseNodeResponse> void buildNodesHeader(final XContentBuilder builder, final Params params,
+                                                                                final BaseNodesResponse<NodeResponse> response)
+        throws IOException {
+        final int successful = response.getNodes().size();
```
I think I missed the discussion but why isn't all this (this method and the next two) part of BaseNodeResponse's toXContent implementation? It can declare an abstract method that the subclasses can override for their own xcontent? We use that pattern pretty frequently with things like the query builders.

jasontedor replied:
Because not all of the inheriting classes implement `ToXContent`.

pickypg replied:
I had initially done that, which led to the comment from @jasontedor:

https://github.com/elastic/elasticsearch/pull/17964#discussion_r61251687

I had initially done the extended approach.

nik9000 replied:
I understand the thing about not all subclasses implementing toXContent and I agree doing it this way is better in that case. If I had to do it myself I'm not sure I'd do it this way but I don't know how I'd do it. In the case that not all subclasses have toXContent then this is fine/good/maybe the best way.

jasontedor replied:
Note that we also use this pattern for share responses which are in the same situation.

nik9000 commented on core/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java:
```diff
@@ -431,6 +431,29 @@ public String getWriteableName() {
+        for (int i = 0; i < expected.size(); ++i) {
+            assertEquals(expected.get(i).value, loaded.get(i).value);
+        }
+    }
```
You can also assert that there are no bytes left to read I believe. Some of the other tests in this file do it and it is a good idea I think.

pickypg replied:
Added.

nik9000 replied:
Did you push the change that added it? I don't see it.

pickypg replied:
I'm letting the tests run first. It takes a _long_ time for it to make it through everything locally.

nik9000 commented:
I left a few comments - mostly small things that have no right to block merging. My only concern is around whether or not we need the new listener - essentially should we just move all of that header logic into the toXContent methods of the BaseNodeResponse?

jasontedor commented:
> My only concern is around whether or not we need the new listener - essentially should we just move all of that header logic into the toXContent methods of the BaseNodeResponse?

BaseNodeResponse doesn't have a `toXContent` method nor should it because it would be a lie because not all of the response objects that inherit from implement `ToXContent`.

nik9000 commented:
> BaseNodeResponse doesn't have a toXContent method nor should it because it would be a lie because not all of the response objects that inherit from implement ToXContent.

Makes sense to me then.

I'm happy with it.

pickypg commented:
@nik9000 Added it now that local tests have finished. I like having the nodes read/written from an abstract method. It also removes a bunch of boilerplate from the child classes.

nik9000 commented:
LGTM. I know there is still discussion but I don't think it is the kind of discussion that blocks merging?

jasontedor commented:
LGTM.
